ecosystem/promise.md

```diff
@@ -0,0 +1,141 @@
+# Purpose
+
+Superchain interop enables async programming at the smart contract level. Onchain state will increasingly become more fragmented as we horizontally scale. Thus accessing remote state as easy and familiar as possible for devs.
+
+## Goals
+
+- Provide a canonical way to query the remote state of any contract
+- Avoid extra contract deployments to faciliate remote queries
+
+## Non-Goals
+
+- Handle dependencies between different cross chain messages.
+
+# Problem Statement + Context
+
+In solidity, if we wanted to query the `balanceOf` function of an ERC20 that lives on a different chain, gateway contracts would have to deployed that could ping-pong the return value back the a handler on the sending chain.
+
+```solidity
+contract ERC20BalanceGateway {
+    L2ToL2CrossDomainMessenger messenger;
+
+    // Gateway contract that can be used to query ERC20s with a specified handler
+    function query(IERC20 token, address account, bytes4 selector) external {
+        uint256 balance = token.balanceOf(account);
+
+        // Message sent back with the value as an encoded argument
+        (address sender, uint256 source) = messenger.crossDomainMessageContext();
+        messenger.sendMessage(source, sender, abi.encode(selector, balance));
+    }
+}
+```
+
+Contracts like the above gateway would have to be created and deployed on every target chain while covering all uses cases for contract reads.
+
+- Deployment overhead for the Gateway. N+1 contract addresses as these kinds of contracts are created.
+
+# Proposed Solution
+
+When a message is relayed the `L2ToL2CrossDomainMessenger` emits:
+
+`RelayedMessage(uint256 indexed source, uint256 indexed messageNonce, bytes32 indexed messageHash);`
+
+By including the return data of the target call, this event itself can be used to continue any attached continuations on the source without a message sent back or additional data propogated from the sending chain.
+
+```solidity
+contract L2ToL2CrossDomainMessenger {
+    function relayMessage(Identifier calldata _id, bytes calldata _sentMessage) {
+        // include return data
+        (, returnData_) = target.call{ value: msg.value }(message);
+        emit RelayedMessage(source, nonce, messageHash, returnData_);
+    }
+}
+```
+
+The returned `messageHash` which identifies a sent messages can be behave like a Promise that the caller can attach continuations to. All state
+related to the callback is managed on the sending chain and works with any remote target.
+
+```solidity
+library Promise {
```

Comment:
I think we should do some thinking around how to make the API for incentivized relay / callbacks as minimal (and flexible) as possible. This makes me wary about using a Library because then it's harder to hide the underlying callback logic and harder to update.

I think it might be easier to use a dedicated contract for registering callbacks that you call vs using a library

Reply:
yea totally agree, the functions here are marked as external meaning this library gets deployed as a dedicated contract

no point in declaring this as a lib though, i'll update this to be a contract